File: source/sparkR/mlp.r
# Databricks notebook source
library(SparkR)

sparkR.session()

#tiempo inicio del algoritmo
timestartINIGlobal <- Sys.time()

##########################################
# Load the data stored as SparkDataframe
##########################################
df <- read.df("FileStore/tables/kdd_training.txt", "csv", header = "true", inferSchema = "true")

##########################################
# Preparación de los datos
##########################################

#tiempo INI preparación de los datos
timestartINIDatos <- Sys.time()

#normal -> 1.0, attack -> 0.0
df <- withColumn(df,"class", ifelse(df$class != "normal", 0.0, 1.0))
df <- drop(df, "is_host_login")
#drop categorical column
df <- drop(df, "service")

#convertir categoricas en numericas
df <- withColumn(df,"protocol_type", ifelse(df$protocol_type == "tcp", 0, ifelse(df$protocol_type == "udp", 1, 2)))
df <- withColumn(df,"flag", ifelse(df$flag == "SF", 0, ifelse(df$flag == "S0", 1, ifelse(df$flag == "S1", 2, ifelse(df$flag == "S2", 3, ifelse(df$flag == "S3", 4, ifelse(df$flag == "SH", 5, ifelse(df$flag == "OTH", 6, ifelse(df$flag == "REJ", 7, ifelse(df$flag == "RSTO", 8, ifelse(df$flag == "RSTOS0", 9, 10)))))))))))

#print(paste(c("distinct ", collect(distinct(select(df, 'protocol_type')))), collapse = " "))
#print(paste(c("distinct ", collect(distinct(select(df, 'service')))), collapse = " "))
#print(paste(c("distinct ", collect(distinct(select(df, 'flag')))), collapse = " "))
print(head(df))
#printSchema(df)
#print(count(df))

splitDF_list <- randomSplit(df, c(0.7, 0.3))
df_training <- splitDF_list[[1]]
df_test <- splitDF_list[[2]]

#tiempo FIN preparación de los datos
timestartFINDatos <- Sys.time()

#tiempo INI generación del modelo
timestartINIModel <- Sys.time()

#2 classes
#count classes y lo convierte a Rdataframe para operar con el, como número de clusters
nm_output <- collect(select(df, countDistinct(df$class)))
nm_input<-ncol(df) - 1 #eliminar la clase

print(paste(c("nm_output ", nm_output), collapse = " "))
print(paste(c("nm_input", nm_input), collapse = " "))

# capas definidas para la red neuronal:
# capa de entrada con nm_input entradas (1 por feature), 
# dos capas intermedias de 10 neuronas
# y una capa de salida con nm_output neuronas (1 por cada clase de salida)
layers = c(nm_input, 10, 10, nm_output)

# Fit a multilayer perceptron classification model with mlp
model <- spark.mlp(df_training, class ~ ., blockSize = 128, layers = layers, maxIter = 100)#, seed=1234

# Model summary
#print(summary(model))

#tiempo FIN generación del modelo
timestartFINModel <- Sys.time()

#tiempo INI generación de las predicciones/validacion del modelo
timestartINITest <- Sys.time()

# Prediction
predictions <- predict(model, df_test)

#printSchema(predictions)
print(head(predictions))

#tiempo FIN generación de las predicciones/validacion del modelo
timestartFINTest <- Sys.time()

#predicciones erróneas
errorPredictions = count(filter(predictions, predictions$prediction != predictions$class))

#predicciones correctas
correctPredictions = count(filter(predictions, predictions$prediction == predictions$class))

total <- count(predictions)
positive = count(filter(predictions, predictions$prediction == 1.0))
negative = count(filter(predictions, predictions$prediction == 0.0))

#ratios de predicciones erróneas y correctas
ratioKO <- errorPredictions / total
ratioOK <- correctPredictions / total

#calculo de true positive, true negative, false positive, false negative
tp = count(filter(filter(predictions, predictions$prediction == 1.0), predictions$class == 1.0))
tn = count(filter(filter(predictions, predictions$prediction == 0.0), predictions$class == 0.0))
fp = count(filter(filter(predictions, predictions$prediction == 1.0), predictions$class == 0.0))
fn = count(filter(filter(predictions, predictions$prediction == 0.0), predictions$class == 1.0))

tp_rate = tp/positive
tn_rate = tn/negative
fp_rate = fp/negative
fn_rate = fn/positive

print(paste(c("total dataset: ", count(df)), collapse = " "))
print(paste(c("total entrenamiento: ", count(df_training)), collapse = " "))
print(paste(c("total validacion: ", count(df_test)), collapse = " "))

print(paste(c("Predictions total: ", total), collapse = " "))
print(paste(c("Correct predictions: ", correctPredictions), collapse = " "))
print(paste(c("Incorrect predictions: ", errorPredictions), collapse = " "))
print(paste(c("Ratio incorrect: ", ratioKO), collapse = " "))
print(paste(c("Ratio correct: ", ratioOK), collapse = " "))

print(paste(c("TP: ",tp_rate), collapse = " "))
print(paste(c("TN: ",tn_rate), collapse = " "))
print(paste(c("FP: ",fp_rate), collapse = " "))
print(paste(c("FN: ",fn_rate), collapse = " "))

print(paste(c("Time taken to prepare the data",difftime(timestartFINDatos, timestartINIDatos, units = "secs")), collapse = " "))
print(paste(c("Time taken to generate the model",difftime(timestartFINModel, timestartINIModel, units = "secs")), collapse = " "))
print(paste(c("Time taken to test the model",difftime(timestartFINTest, timestartINITest, units = "secs")), collapse = " "))
print(paste(c("Time taken to run the algorithm",difftime(Sys.time(), timestartINIGlobal, units = "secs")), collapse = " "))

File: source/sparkML/mlp.py
import sys
import os
from pyspark import SparkContext, SQLContext
from pyspark.ml import Pipeline
from pyspark.ml.classification import MultilayerPerceptronClassifier
from pyspark.ml.evaluation import MulticlassClassificationEvaluator
from pyspark.ml.feature import StringIndexer, OneHotEncoder,VectorIndexer, VectorAssembler

import datetime

from pyspark.sql import SparkSession
from pyspark.ml.linalg import Vectors, VectorUDT

if __name__ == "__main__":
    spark = SparkSession\
        .builder\
        .appName("Multilayer Perceptron")\
        .getOrCreate()

#tiempo preparación de los datos
timestartGlobal = datetime.datetime.now()

# Load the data stored as dataframe
df = spark.read.format("com.databricks.spark.csv").option("header", "true").option("inferSchema", "true").load("/FileStore/tables/kdd_training.txt")

df.cache()

##########################################
#preparación de los datos
# fichero tipo label y features
##########################################

#tiempo preparación de los datos
timestartDatos = datetime.datetime.now()

from pyspark.sql import functions as F
#solo 2 clases, normal y attack (todos los que no son normal)
df = df.withColumn("class", F.when(df["class"]!="normal","attack").otherwise(df["class"]))

#df.printSchema()

#categoricas
categorical_columns = ["protocol_type","service","flag","land","logged_in","is_guest_login"]

#numericas = todas las columnas menos las categoricas (incluyendo el class)
numeric_columns = [col for col in df.columns if col not in categorical_columns]

#eliminamos la columna class-label
numeric_columns = [c for c in numeric_columns if c not in ["class","is_host_login"]]
#print(numeric_columns)

# StringIndexer para las variables categóricas
stringindexer_stages = [StringIndexer(inputCol=c, outputCol='strindexed_' + c) for c in categorical_columns]

# los nombres de los indices de las categóricas para añadir a las features
stringindexer_columns = ['strindexed_' + c for c in categorical_columns]

#vector de stringIndexer, utilizando OneHotEncoder
#onehotencoder_stages = [OneHotEncoder(inputCol='strindexed_' + c, outputCol='onehotencoder_' + c) for c in categorical_columns]

#onehotencoder_columns = ['onehotencoder_' + c for c in categorical_columns]

# StringIndexer para la columna class-label
labelindexer_stages = [StringIndexer(inputCol='class', outputCol='label')]

# feature columns (numericas + categoricas_indexed) para el vectorAssembler de features
feature_columns = numeric_columns + stringindexer_columns

vectorassembler_stage = VectorAssembler(inputCols=feature_columns, outputCol='features') 

# para el pipeline del dataframe inicial
all_stages = stringindexer_stages + [vectorassembler_stage] + labelindexer_stages

pipeline_data = Pipeline(stages=all_stages)

pipeline_model = pipeline_data.fit(df)

# transform the data
final_columns = ['features', 'label']

data_df = pipeline_model.transform(df).select(final_columns)

#2 classes
df_output = df.groupBy("class").count()

df_output.show(2)
nm_output = df_output.count()

print("nm_output " + str(nm_output))

nm_input = len(numeric_columns)+len(categorical_columns)
print("nm_input " + str(nm_input))

#tiempo fin de preparacion de datos
timeendDatos = datetime.datetime.now()

data_df.show(10)
##########################################
# FIN preparación de los datos
# fichero tipo label y features
##########################################

##########################################
# Generación del modelo
##########################################

#tiempo generación del modelo
timestartModel = datetime.datetime.now()

# Dividir el dataset en datos de entrenamiento (70%) y datos para validacion del modelo (30%)
(trainingData, testData) = data_df.randomSplit([0.7, 0.3])

trainingData.cache()
testData.cache()

print(data_df.count())
print(trainingData.count())
print(testData.count())

# capas definidas para la red neuronal:
# capa de entrada con nm_input entradas (1 por feature), 
# dos capas intermedias de 10 neuronas
# y una capa de salida con nm_output neuronas (1 por cada clase de salida)
layers = [nm_input, 10, 10, nm_output]

# create the trainer and set its parameters
trainer = MultilayerPerceptronClassifier(maxIter=100, layers=layers, blockSize=128)

# train the model
model = trainer.fit(trainingData)

#tiempo generación del modelo
timeendModel = datetime.datetime.now()

#tiempo validacion del modelo
timestartTest = datetime.datetime.now()

predictions = model.transform(testData)

predictions.show(10)

predictionAndLabels = predictions.select("prediction", "label")
evaluator = MulticlassClassificationEvaluator(metricName="accuracy")
accuracy = evaluator.evaluate(predictionAndLabels)

#tiempo validación del modelo
timeendTest = datetime.datetime.now()

#predictions.show(10)
#predictions.groupBy("prediction").count().show(40)

#número de predicciones total
total = float(predictions.count())

#predicciones erróneas
errorPredictions = predictions.filter(predictions['prediction'] != predictions['label']).count()
#predicciones correctas
correctPredictions = predictions.filter(predictions['prediction'] == predictions['label']).count()

#ratios de predicciones erróneas y correctas
ratioKO = errorPredictions / total
ratioOK = correctPredictions / total

#calculo de true positive, true negative, false positive, false negative
tp = (predictions.filter(predictions['prediction'] == 1.0).filter(predictions['label'] == 1.0).count())/total
tn = (predictions.filter(predictions['prediction'] == 0.0).filter(predictions['label'] == 0.0).count())/total
fp = predictions.filter(predictions['prediction'] == 1.0).filter(predictions['label'] == 0.0).count()/total
fn = predictions.filter(predictions['prediction'] == 0.0).filter(predictions['label'] == 1.0).count()/total


print("Exactitud (accuracy) = %g " % accuracy)
print("Error rate = %g " % (1.0 - accuracy))
print("Predictions total: " + str(total))
print("Correct predictions: " + str(correctPredictions))
print("Incorrect predictions: " + str(errorPredictions))
print("Ratio incorrect: " + str(ratioKO))
print("Ratio correct: " + str(ratioOK))
print("TP: " + str(tp))
print("TN: " + str(tn))
print("FP: " + str(fp))
print("FN: " + str(fn))
    
timeendGlobal = datetime.datetime.now()

print "Time taken to prepare data: " + str(round((timeendDatos-timestartDatos).total_seconds(), 2)) + " seconds";
print "Time taken to generate the model: " + str(round((timeendModel-timestartModel).total_seconds(), 2)) + " seconds";
print "Time taken to validate the model: " + str(round((timeendTest-timestartTest).total_seconds(), 2)) + " seconds";
print "Time taken to run the algorithm: " + str(round((timeendGlobal-timestartGlobal).total_seconds(), 2)) + " seconds";
